From: Nathan D. Fontenot <nfont@austin.ibm.com>
Subject: Mark preferred offline state as CPU_STATE_OFFLINE before removing cpus
Patch-Mainline: Not yet
References: bnc#590154

The preferred offline state needs to be marked as CPU_STATE_OFFLINE before
online cpus are hotplug removed.
The existing behavior can cause the dlpar operation to fail if the preferred
offline state is set to CPU_STATE_INACTIVE when hotplug removing the cpu
since this causes the cpu to not be completely removed from the system.

Signed-off-by: Thomas Renninger <trenn@suse.de>

---
 arch/powerpc/platforms/pseries/dlpar.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

Index: linux-3.0-tmp-jikos/arch/powerpc/platforms/pseries/dlpar.c
===================================================================
--- linux-3.0-tmp-jikos.orig/arch/powerpc/platforms/pseries/dlpar.c
+++ linux-3.0-tmp-jikos/arch/powerpc/platforms/pseries/dlpar.c
@@ -477,6 +477,8 @@ static int dlpar_offline_cpu(struct devi
 			if (get_cpu_current_state(cpu) == CPU_STATE_OFFLINE)
 				break;
 
+			set_preferred_offline_state(cpu, CPU_STATE_OFFLINE);
+
 			if (get_cpu_current_state(cpu) == CPU_STATE_ONLINE) {
 				set_preferred_offline_state(cpu, CPU_STATE_OFFLINE);
 				cpu_maps_update_done();
@@ -492,7 +494,6 @@ static int dlpar_offline_cpu(struct devi
 			 * The cpu is in CPU_STATE_INACTIVE.
 			 * Upgrade it's state to CPU_STATE_OFFLINE.
 			 */
-			set_preferred_offline_state(cpu, CPU_STATE_OFFLINE);
 			BUG_ON(plpar_hcall_norets(H_PROD, intserv[i])
 								!= H_SUCCESS);
 			__cpu_die(cpu);
